### PR TITLE
Report enabled/disabled plans/tests/stories via option

### PR DESCRIPTION
**Pull request by idorax**
fix: #1391

### CODE REVIEW

**Comment by idorax** [comment]
With the patch, I manually tested it, the test results look good.

##### 01 - FMF files for testing
```bash
(tmt) huanli@kvm-01-guest02:foo$ tree -a
.
├── .fmf
│   └── version
└── plan
    ├── bar.fmf
    └── foo.fmf

2 directories, 3 files
(tmt) huanli@kvm-01-guest02:foo$ 
(tmt) huanli@kvm-01-guest02:foo$ cat plan/foo.fmf
execute:
  how: tmt
enabled: false
(tmt) huanli@kvm-01-guest02:foo$ 
(tmt) huanli@kvm-01-guest02:foo$ cat plan/bar.fmf
execute:
  how: tmt
```

##### 02 - Test result of `tmt plans ls`

```bash
(tmt) huanli@kvm-01-guest02:foo$ tmt plans ls --help
Usage: tmt plans ls [OPTIONS] [REGEXP|.]

  List available plans.

  Regular expression can be used to filter plans by name. Use '.' to select
  plans under the current working directory.

Options:
  -a, --all               List all plans even if property enabled is false.
  -f, --filter FILTER     Apply advanced filter (see 'pydoc fmf.filter').
...<snip>...................................................................
  -q, --quiet             Be quiet. Exit code is just enough for me.
  --help                  Show this message and exit.
(tmt) huanli@kvm-01-guest02:foo$ 
(tmt) huanli@kvm-01-guest02:foo$ tmt plans ls
/plan/bar
(tmt) huanli@kvm-01-guest02:foo$ 
(tmt) huanli@kvm-01-guest02:foo$ tmt plans ls --all
/plan/bar
/plan/foo
(tmt) huanli@kvm-01-guest02:foo$ 
(tmt) huanli@kvm-01-guest02:foo$ tmt -vvv plans ls -a
/plan/bar
/plan/foo
(tmt) huanli@kvm-01-guest02:foo$ 
```

##### 03 - Test result of `tmt plans show`

```bash
(tmt) huanli@kvm-01-guest02:foo$ tmt plans show --help
Usage: tmt plans show [OPTIONS] [REGEXP|.]

  Show plan details.

  Regular expression can be used to filter plans by name. Use '.' to select
  plans under the current working directory.

Options:
  -a, --all               Show all plans even if property enabled is false.
  -f, --filter FILTER     Apply advanced filter (see 'pydoc fmf.filter').
...<snip>...................................................................
  -q, --quiet             Be quiet. Exit code is just enough for me.
  --help                  Show this message and exit.
(tmt) huanli@kvm-01-guest02:foo$ tmt plans show
/plan/bar
     enabled true

(tmt) huanli@kvm-01-guest02:foo$ 
(tmt) huanli@kvm-01-guest02:foo$ tmt plans show --all
/plan/bar
     enabled true

/plan/foo
     enabled false

(tmt) huanli@kvm-01-guest02:foo$ tmt -vvv plans show -a
/plan/bar
    discover 
         how shell
   provision 
         how virtual
       image fedora
        user root
      memory 2048
        disk 10
  connection session
        arch x86_64
     prepare 
         how shell
      script 
     execute 
         how tmt
      report 
         how display
      finish 
         how shell
      script 
     enabled true

/plan/foo
    discover 
         how shell
   provision 
         how virtual
       image fedora
        user root
      memory 2048
        disk 10
  connection session
        arch x86_64
     prepare 
         how shell
      script 
     execute 
         how tmt
      report 
         how display
      finish 
         how shell
      script 
     enabled false

(tmt) huanli@kvm-01-guest02:foo$ 
```

The output of `tmt plans ls` is expected by issue #1391,  but I'm not sure the output of `tmt plans show` is expected.  @lukaszachy and @sopos, please help to review, thanks!

**Comment by idorax** [comment]
I added option `--enabled` and `--disabled` to the `name_filter_condition` decorator according to the agreement from the latest discussion. @psss, @lukaszachy and @sopos, please help to review, thanks!

**Comment by sopos** [comment]
looks good to me but do not take me as a relevant authority

**Review by psss — CHANGES_REQUESTED** [review]
Thanks for working on this. Please, include at least a basic test coverage for the new functionality.

**Comment by idorax** [comment]
> Thanks for working on this. Please, include at least a basic test coverage for the new functionality.

Hi @psss, the related tests are added. Please help to review, thanks!

BTW, do you know why `stories_ls()` and `stories_show()` uses decorator `name_filter_condition_long` but `plans_ls`, `plans_show()`, `tests_ls()` and `tests_show` uses decorator `name_filter_condition`? e.g.
* `stories_ls`: https://github.com/teemtee/tmt/blob/main/tmt/cli.py#L897
* `plans_ls`: https://github.com/teemtee/tmt/blob/main/tmt/cli.py#L731

**Comment by psss** [comment]
> BTW, do you know why `stories_ls()` and `stories_show()` use decorator `name_filter_condition_long`...

Yes, originally the `stories` subcommand used both short and long options, for example `-c` and `--covered`, which would collide with `-c` as `--condition` or `-v` as `--verbose`.

File: tmt/base.py [hunk 1]
```diff
@@ -1435,6 +1435,22 @@ def name_filter(nodes):
                 node for node in nodes
                 if any([re.search(name, node.name) for name in cmd_line_names])]
 
+        # Get tests whose property 'enabled' is true (when option --enabled is specified)
+        if Test._opt('enabled'):
```
Comment (lukaszachy) [inline tmt/base.py:1439]: I think this doesn't work as intended. Filters is a  list however each item may contain more values joined by `&` or `|`.

So let's say I execute `tmt t ls --filter 'test:echo&enabled:true'` then `filters=['test:echo&enabled:true']` is used. So for `tmt t ls --filter 'test:echo&enabled:true' --disabled` it becomes `filters=['test:echo&enabled:true', 'enabled:false']` which obviously prints nothing as my cmdline 'enabled:true' was not removed.

If the preferred way to achieve disable/enable is using filter then IMO tmt needs to remove any `enabled:XXX` regardless the value (there is '-' for not so "enabled:-false" is valid and working) and remove & or | used to for "enabled" as well. IIRC we allow anything in the value, but please confirm in fmf.filter implementation, so value is either until the end of the string or & or |. This removal could cause empty item in filters so that should be addressed too. 

So for my `filters=['test:echo&enabled:true']` it would remove enabled to get `filters=['test:echo]` (& has to be removed as well) and later add new item so `filters=['test:echo', 'enabled:false']` is the final filter.

We don't have priorities for AND and OR so just remove that character before 'enabled:' from the filter string and it should be OK.


Also would be best to have single place for this filter transformation method and use it from all places (easier to test via unit test etc.)

**Review by lukaszachy — CHANGES_REQUESTED** [review]
Existing test in beakerlib is good. 
'filters' manipulation should be its own function, with unit test for it (including corner cases).

File: tmt/base.py (hunk 1 above)
Comment (idorax) [reply tmt/base.py:1439]: To handle both `&`  and `|`, looks it is a bit complicated to remove `&, |` from `filters=['test:echo&enabled:true']` or `filters=['test:echo|enabled:true']`. Currently `&` and `|` is not handled in `Tree.plans()`, `Tree.tests()` and `Tree.stories()`. e.g.
```python
$ python3 -m pdb ~/.virtualenvs/tmt/bin/tmt \
   tests show --filter 'enabled:true|test:echo'
(Pdb) b tmt/base.py:1439
Breakpoint 1 at /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/base.py:1439
(Pdb) c
> /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/base.py(1439)tests()
-> if Test._opt('enabled'):
(Pdb) l
1434 	            return [
1435 	                node for node in nodes
1436 	                if any([re.search(name, node.name) for name in cmd_line_names])]
1437 	
1438 	        # Get tests whose property 'enabled' is true (when option --enabled is specified)
1439B->	        if Test._opt('enabled'):
1440 	            # '--enabled' win '--filter enabled:false'
1441 	            while 'enabled:false' in filters:
1442 	                filters.remove('enabled:false')
1443 	            if 'enabled:true' not in filters:
1444 	                filters.append('enabled:true')
(Pdb) p filters
['enabled:true|test:echo']
(Pdb) c
/tests/disabled2
     summary Test tmt tests ls/show --disabled
        test echo
        path /tests
      manual false
    duration 5m
     enabled false
      result respect

/tests/enabled1
     summary Test tmt tests ls/show --enabled
        test /bin/true
        path /tests
      manual false
    duration 5m
     enabled true
      result respect

/tests/enabled2
     summary Test tmt tests ls/show --enabled
        test echo
        path /tests
      manual false
    duration 5m
     enabled true
      result respect

The program exited via sys.exit(). Exit status: 0
```

Hence, we can use `re.sub()` to replace those filters including `enabled:(true|false)`. e.g.
```python
$ cat -n /tmp/foo.py 
     1	#!/usr/bin/python3
     2	import sys
     3	import re
     4	
     5	
     6	def revise_filters(filters):
     7	    out = []
     8	    for filter in filters:
     9	        if 'enabled:' in filter:
    10	            filter = re.sub(r'enabled:(true|false)', 'enabled:true', filter)
    11	        out.append(filter)
    12	    return out
    13	
    14	
    15	def main(argc, argv):
    16	    print(revise_filters(argv[1:]))
    17	    return 0
    18	
    19	
    20	if __name__ == '__main__':
    21	    sys.exit(main(len(sys.argv), sys.argv))
$ 
$ /tmp/foo.py 'test:echo&enabled:false'
['test:echo&enabled:true']
$ /tmp/foo.py 'enabled:false|test:echo'
['enabled:true|test:echo']
``` 

@lukaszachy, what do you think?

**Comment by lukaszachy** [comment]
@idorax If we are OK with returning [] when user's filter and --disabled/--enabled conflict then you can do it in by 
```
if Test._opt('enabled'):
  post_filter = ['enabled:true']
elif Test._opt('disabled'):
  post_filter = ['enabled:false']
else:
  post_filter = []
self._filters_conditions(tests, filters + post_filters, conditions, links, excludes)
```

no need to call _filters_conditions twice

**Comment by idorax** [comment]
> @idorax If we are OK with returning [] when user's filter and --disabled/--enabled conflict then you can do it in by
> ...
> no need to call _filters_conditions twice

I think it's acceptable to return [] when user's filter and --disabled/--enabled conflict because if user uses filter, he should be able to know how to handle `--filter` with `--enabled|--disabled`.

I have updated the related `tmt/base.py` according to your suggestion, please help to review again, thanks!

File: tmt/base.py (hunk 1 above)
Comment (psss) [reply tmt/base.py:1439]: Agreed that it's not necessary to handle the conflicting `--filter` content with `--enabled` and `--disabled`. Let's just append the additional filter.

File: tmt/base.py [hunk 2]
```diff
@@ -1499,6 +1515,14 @@ def stories(self, keys=None, names=None, filters=None, conditions=None,
         links = (links or []) + list(Story._opt('links', []))
         excludes = (excludes or []) + list(Story._opt('exclude', []))
 
+        # Get post filter to support option '--enabled/--disabled'
+        if Story._opt('enabled'):
+            post_filters = ['enabled:true']
```
Comment (psss) [inline tmt/base.py:1520]: What about just extending the `filters` list?

```suggestion
            filters.append('enabled:true')
```

There's no additional processing of it so this will work nicely and we can also get rid of the `else` part below.

File: tmt/cli.py [hunk 3]
```diff
@@ -112,6 +112,12 @@ def name_filter_condition(function: Callable[..., Any]) -> Callable[..., Any]:
         click.option(
             '-c', '--condition', 'conditions', metavar="EXPR", multiple=True,
             help="Use arbitrary Python expression for filtering."),
+        click.option(
+            '--enabled', is_flag=True,
+            help="Show enabled plans/tests/stories."),
```
Comment (psss) [inline tmt/cli.py:117]:
```suggestion
            help="Show only enabled tests, plans or stories."),
```

File: tmt/cli.py [hunk 4]
```diff
@@ -112,6 +112,12 @@ def name_filter_condition(function: Callable[..., Any]) -> Callable[..., Any]:
...
+        click.option(
+            '--disabled', is_flag=True,
+            help="Show disabled plans/tests/stories."),
```
Comment (psss) [inline tmt/cli.py:120]:
```suggestion
            help="Show only disabled tests, plans or stories."),
```

**Review by psss — CHANGES_REQUESTED** [review]
Thanks for implementing this! Looks good to me. Proposing just a couple of minor improvements.

**Review by lukaszachy — APPROVED** [review]
LGTM. I have no preference for the proposed 'filters' change.

File: tmt/cli.py (hunk 3 above)
Comment (idorax) [reply tmt/cli.py:117]: Updated.

File: tmt/cli.py (hunk 4 above)
Comment (idorax) [reply tmt/cli.py:120]: Updated.

File: tmt/base.py (hunk 2 above)
Comment (idorax) [reply tmt/base.py:1520]: @psss and @lukaszachy, `post_filters` is removed and use `filters.append()` instead.

Comment (psss) [reply tmt/base.py:1520]: Thanks!

File: tests/plan/show/test.sh [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+. /usr/share/beakerlib/beakerlib.sh || exit 1
+
+rlJournalStart
+    rlPhaseStartSetup
+        rlRun "pushd data"
+        rlRun "set -o pipefail"
+    rlPhaseEnd
+
+    rlPhaseStartTest
+        rlRun -s "tmt plans ls" 0 "List all plans by default"
```
[the remark this answers is not in the document]
Comment (idorax) [reply tests/plan/show/test.sh]: Sure, the related 3 test.sh files are updated.

**Review by psss — APPROVED** [review]
Thanks for all the improvements. Now looks good!
